File: app/src/main/java/com/chinachino/mvvm/Repositories/MovieRepository.java
package com.chinachino.mvvm.Repositories;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.chinachino.mvvm.Requests.MovieAPIClient;
import com.chinachino.mvvm.models.Details;
import com.chinachino.mvvm.models.Result;

import java.util.List;

public class MovieRepository {
    private static MovieRepository instance;
    private MovieAPIClient movieAPIClient;
    private String query;
    private int pageNumber;

    public static MovieRepository getInstance() {
        if (instance == null) {
            instance = new MovieRepository();
        }
        return instance;
    }

    private MovieRepository() {
        movieAPIClient = MovieAPIClient.getInstance();
    }

    public LiveData<List<Result>> getMovies() {
        return movieAPIClient.getMovies();
    }
    public LiveData<Details> getMovieDetails(){
        return movieAPIClient.getMovieDetails();
    }
    public void SearchMovieAPI(String query,int page){
        this.query = query;
        this.pageNumber = page;
        movieAPIClient.SearchMovieAPI(query,page);
    }
    public void SearchMovieDetails(int movieID){
        movieAPIClient.SearchMovieID(movieID);
    }
    public void SearchNextQuery(){
        movieAPIClient.SearchMovieAPI(this.query,this.pageNumber+1);
    }
}

File: app/src/main/java/com/chinachino/mvvm/repositories/MovieRepository.kt
package com.chinachino.mvvm.repositories

import androidx.lifecycle.LiveData
import com.chinachino.mvvm.requests.MovieAPIClient
import com.chinachino.mvvm.models.Details
import com.chinachino.mvvm.models.Result

object MovieRepository  {
    private val movieAPIClient: MovieAPIClient = MovieAPIClient()
    private var  query = ""
    private var pageNumber = 0
    val movies: LiveData<List<Result>>
        get() = movieAPIClient.movies
    val movieDetails: LiveData<Details?>
        get() = movieAPIClient.movieDetails

    fun searchMovieAPI(query: String, page: Int) {
        this.query = query
        pageNumber = page
        movieAPIClient.searchMovieAPI(query, page)
    }

    fun searchMovieDetails(movieID: Int) {
        movieAPIClient.searchMovieID(movieID)
    }

    fun searchNextQuery() {
        movieAPIClient.searchMovieAPI(query, pageNumber + 1)
    }

    val isRequestTimedOut: LiveData<Boolean>
        get() = movieAPIClient.isRequestTimeOut()



}